pkg/ipcache/ipcache.go

```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+				// Skip calling back the listeners, since the endpoint IP has
+				// precedence over the new CIDR.
+				callbackListeners = false
+			}
+		}
+	} else if endpointIP := net.ParseIP(ip); endpointIP != nil { // Endpoint IP.
+		// Convert the endpoint IP into an equivalent full CIDR.
+		bits := net.IPv6len * 8
+		if endpointIP.To4() != nil {
+			bits = net.IPv4len * 8
+		}
+		cidr = &net.IPNet{
+			IP:   endpointIP,
+			Mask: net.CIDRMask(bits, bits),
+		}
+
+		// Check whether the upserted endpoint IP will shadow that CIDR, and
+		// replace its mapping with the listeners if that was the case.
+		if !found {
+			if cidrIdentity, cidrFound := ipc.ipToIdentityCache[cidr.String()]; cidrFound {
+				if cidrIdentity.ID != newIdentity.ID {
+					scopedLog.Debug("New endpoint IP started shadowing existing CIDR to identity mapping")
+					oldIdentity = &cidrIdentity.ID
+				} else {
+					// The endpoint IP and the CIDR are associated with the
+					// same identity. Nothing changes for the listeners.
+					callbackListeners = false
+				}
+			}
+		}
+	} else {
+		scopedLog.Debug("Attempt to upsert invalid IP into ipcache layer")
+		return false
+	}
+
+	scopedLog.Debug("Upserting IP into ipcache layer")
+
+	// Update both maps.
+	ipc.ipToIdentityCache[ip] = newIdentity
+	// Delete the old identity, if any.
+	if found {
+		delete(ipc.identityToIPCache[cachedIdentity.ID], ip)
+		if len(ipc.identityToIPCache[cachedIdentity.ID]) == 0 {
+			delete(ipc.identityToIPCache, cachedIdentity.ID)
+		}
+	}
+	if _, ok := ipc.identityToIPCache[newIdentity.ID]; !ok {
+		ipc.identityToIPCache[newIdentity.ID] = map[string]struct{}{}
+	}
+	ipc.identityToIPCache[newIdentity.ID][ip] = struct{}{}
+
+	if callbackListeners {
+		for _, listener := range ipc.listeners {
+			listener.OnIPIdentityCacheChange(Upsert, *cidr, hostIP, oldIdentity, newIdentity.ID)
+		}
 	}
 
 	return true
 }
 
 // deleteLocked removes removes the provided IP-to-security-identity mapping
 // from ipc with the assumption that the IPCache's mutex is held.
-func (ipc *IPCache) deleteLocked(IP string) {
-	log.WithFields(logrus.Fields{
-		logfields.IPAddr: IP,
-	}).Debug("Removing from ipcache layer")
+func (ipc *IPCache) deleteLocked(ip string) {
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.IPAddr: ip,
+	})
 
-	identity, found := ipc.ipToIdentityCache[IP]
-	if found {
-		delete(ipc.ipToIdentityCache, IP)
-		delete(ipc.identityToIPCache[identity.ID], IP)
-		if len(ipc.identityToIPCache[identity.ID]) == 0 {
-			delete(ipc.identityToIPCache, identity.ID)
-		}
+	cachedIdentity, found := ipc.ipToIdentityCache[ip]
+	if !found {
+		scopedLog.Debug("Attempt to remove non-existing IP from ipcache layer")
+		return
+	}
+
+	var cidr *net.IPNet
+	cacheModification := Delete
+	var oldIdentity *identity.NumericIdentity
+	newIdentity := cachedIdentity
+	callbackListeners := true
 
+	var err error
+	if _, cidr, err = net.ParseCIDR(ip); err == nil {
```

Comment:
This logic (the next ~40-50LoC) feels like it should be able to be shared with `Upsert()`.